Userland/Applications/Assistant/FuzzyMatch.cpp

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright (c) 2021, Spencer Dixon <spencercdixon@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "FuzzyMatch.h"
+
+#include <stdint.h>
+#include <string.h>
+#include <ctype.h>
+
+// NOTE: This fuzzy match algorithm is taken from here: https://www.forrestthewoods.com/blog/reverse_engineering_sublime_texts_fuzzy_match/
+//       and tuned to work well given the types of results that Assistant returns.
+
+namespace Assistant {
+
+static bool fuzzy_match_recursive(const char* pattern, const char* str, int& outScore, const char* strBegin,
+    uint8_t const* srcMatches, uint8_t* newMatches, int maxMatches, int nextMatch,
+    int& recursionCount, int recursionLimit);
```

Comment:
Please don't use camelCase for variable names.
also, this function is pretty damn chunky.
also, please prefer `StringView` over `const char*` where possible.
also, most of these `int`'s are actually indices/counts, so please use `size_t` for those.
also, please prefer `uN` over `uintN_t` (e.g. `u8` here).
also, `matches` + `maxMatches` sounds like it could be a `Span<u8>`.

I'm not sure what the algorithm is, but it looks to me like you're using `strBegin + some_scan_index` as the string, so perhaps switch to a `StringView` plus an index instead?

Reply:
So yeah wasn't sure what to do about this. The fuzzy algorithm is taken from [here](https://www.forrestthewoods.com/blog/reverse_engineering_sublime_texts_fuzzy_match/), you can see my note [here](https://github.com/SerenityOS/serenity/pull/8266/files/eda5de6c68d32d433bc7a2acd61ab22d6cac2cb4..6610d1932826a211cae869fad6b26a1e529fee75#diff-5816677caa62899602186d3731f71db96d6fc232f6dbb83fe9f3cd2e591e660fR13)

I intentionally preserved the code from the article but I'm happy to port it into a 'SerenityOS' style if that is preferred. Wasn't sure if there was value in leaving it the way it was originally written or not.

Reply:
Importing code isn't really something we do here, it's definitely alright to look at other codebases and cite them as a source, but just copy-paste isn't really an option.

Reply:
Yup fair enough, chatted with @awesomekling about that. Just wrote my own version of the fuzzy matching algorithm using AK datastructures/idioms